<!DOCTYPE html>
<html>
<head>
<title>Exceptions</title>
<meta name="generator" content="Bluefish 2.2.10" >
<meta name="copyright" content="">
<meta name="keywords" content="">
<meta name="description" content="">
<meta name="ROBOTS" content="NOINDEX, NOFOLLOW">
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta http-equiv="content-type" content="application/xhtml+xml; charset=UTF-8">
<meta http-equiv="content-style-type" content="text/css">
<meta http-equiv="expires" content="0">
<link href="yaooc.css" rel="stylesheet" type="text/css">
<script type="text/javascript" src="yaooc.js"></script>
</head>
<body>
<div id="wrapper">
<script>header();</script>
<script>menu();</script>
<div id="content" class="content">
<div style="overflow-y: scroll; height:900px">
<h2>Exceptions</h2>
<p>
Another well known feature of C++ is exceptions. YAOOC has this feature as well.  The syntax is:  
</p>
<pre><code>
  TRY {
  } CATCH (exception_type,exception_var) {
  } ETRY
</code></pre>
<p>
Where exception_type is the type of exception and exception_var is the variable to which the exception is assigned.
Note the comma between exception type and exception var.
</p>
<ul>
<li>
All exceptions must be a descendant of yaooc_exeception (shown below).
</li>
<li>
There can be multiple CATCH blocks.
</li>
<li>
TRY/CATCH blocks can be nested -- either in TRY portion or CATCH portion.
</li>
<li>
CATCH will catch exception of the type specified and types that are derived from exception_type.
</li>
<li>
An exception stack is created for each thread.
</li>
</ul>
<p>
To throw an exception, use the THROW macro.
</p>
<pre><code>
  THROW(new(exception_type))
</code></pre>
<p>
If any thread throws an exception that is not caught, then a message indicating uncaught exception
is printed to stderr and the program is terminated.
</p>
<p>
All exceptions should be a descendant of yaooc_exception.  The definition of yaooc_exception is:
</p>
<pre><code>
  class exception : object {
    table:
      const char* what() const;
  };
</code></pre>
<h3>TRY/CATCH blocks have a pointer bag</h3>
<p>
Objects that were created within the scope of the TRY/CATCH block need to be deleted -- even if an exception is thrown.
YAOOC facilitates this by providing a pointer bag in the TRY/CATCH block. The following macros can be used within a TRY/CATCH block.
</p>

<div class="divTable">
<div class="divTableBody">
<div class="divTableRow">
<div class="divTableHead">Macro</div>
<div class="divTableHead">Purpose</div>
</div>
<div class="divTableRow">
<div class="divTableCell">EPB_SAVE(ptr)</div>
<div class="divTableCell">Save pointer in exception pointer bag and returns value being saved</div>
</div>
<div class="divTableRow">
<div class="divTableCell">EPB_CLEAR</div>
<div class="divTableCell">Removes pointers from exception pointer bag without deleting pointers</div>
</div>
<div class="divTableRow">
<div class="divTableCell">EPB_DELETE</div>
<div class="divTableCell">Deletes pointers in exception pointer bag.  More pointer can still be saved</div>
</div>
</div>
</div>
<p>
Usually, only the EPB_SAVE macro is used.
</p>
<p>
The TRY/CATCH block pointer bag is created when the TRY keyword is encountered.  A unique pointer bag is created for each
TRY encountered.  Pointers saved (EPB_SAVE) in pointer bag will be saved in the current TRY/CATCH block pointer bag. 
The pointer bag is deleted (and all pointers contained) when the TRY/CATCH block goes out of scope. If a method is called within
the TRY/CATCH block, that method can use EPB_SAVE to save pointers in the pointer bag.
</p>
<pre><code>
  TRY {  /* Pointer bag 1 is created here */
    ... /* EPB_SAVE saves to pointer bag 1 */
    TRY { /* Pointer bag 2 is created here */
      ...   /* EPB_SAVE saves to pointer bag 2 */
    } CATCH(exception,e) {
      ...
    } ETRY /* Pointer bag 2 is deleted here */
  } CATCH(exception,e) {
     ...
  } ETRY /* Pointer bag 1 is deleted here */  
</code></pre>
</div>
</div>
</div>
</body>
</html>